File: 038_RemoveNthFromEnd.cpp
// Topic: Remove Nth Node From End of List
// Description: Delete node located n positions from list end.
// Approach: Two-pointer with gap n, remove target in one pass.

#include <bits/stdc++.h>
using namespace std;

struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* removeNthFromEnd(ListNode* head, int n)
{
    ListNode dummy(0);
    dummy.next = head;
    ListNode *first = &dummy, *second = &dummy;
    for (int i = 0; i <= n; ++i) first = first->next;
    while (first)
    {
        first = first->next;
        second = second->next;
    }
    ListNode* todel = second->next;
    second->next = second->next->next;
    delete todel;
    return dummy.next;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    // Test 1
    {
        ListNode* h = new ListNode(1);
        h->next = new ListNode(2);
        h->next->next = new ListNode(3);
        h->next->next->next = new ListNode(4);
        h->next->next->next->next = new ListNode(5);
        h = removeNthFromEnd(h, 2);
        for (ListNode* p = h; p; p = p->next)
        {
            cout << p->val;
            if (p->next)
                cout << " ";
        }
        cout << "\n";
    }
    // Test 2
    {
        ListNode* h = new ListNode(1);
        h = removeNthFromEnd(h, 1);
        if (!h)
            cout << "Empty\n";
        else
        {
            for (ListNode* p = h; p; p = p->next)
            {
                cout << p->val;
                if (p->next)
                    cout << " ";
            }
            cout << "\n";
        }
    }
}

/*
Time Complexity: O(n)
Space Complexity: O(1)
*/

File: 038_RemoveNthFromEnd.py
# Topic: Remove Nth Node From End of List
# Description: Delete node located n positions from list end.
# Approach: Two-pointer with gap n, remove target in one pass.

from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        first = second = dummy
        for _ in range(n + 1):
            first = first.next
        while first:
            first = first.next
            second = second.next
        second.next = second.next.next
        return dummy.next


def print_list(head):
    if not head:
        print("Empty")
        return
    vals = []
    while head:
        vals.append(str(head.val))
        head = head.next
    print(' '.join(vals))


if __name__ == "__main__":
    # Test 1
    h = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    result = Solution().removeNthFromEnd(h, 2)
    print_list(result)

    # Test 2
    h = ListNode(1)
    result = Solution().removeNthFromEnd(h, 1)
    print_list(result)

"""
Time Complexity: O(n)
Space Complexity: O(1)
"""
